docs/madr/decisions/018-fault-injection-policy.md

```diff
@@ -0,0 +1,230 @@
+resource for managing fault injection.
+
+FaultInjection in Kuma is implemented with Envoy's [fault injection
+support](https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/fault_filter).
+
+## Considered Options
+
+* Create a MeshFaultInjection
+
+## Decision Outcome
+
+Chosen option: create a MeshFaultInjection
+
+## Decision Drivers
+
+- Replace existing policies with new policy-matching compliant resources
+
+## Solution
+
+### Current configuration
+Below is a sample fault injection configuration.
+
+```yaml
+spec:
+  sources:
+    - match:
+        kuma.io/service: frontend_default_svc_80
+        kuma.io/protocol: http
+  destinations:
+    - match:
+        kuma.io/service: backend_default_svc_80
+        kuma.io/protocol: http
+  conf:
+    abort:
+      httpStatus: 500
+      percentage: 50
+    delay:
+      percentage: 50.5
+      value: 5s
+    responseBandwidth:
+      limit: 50 mbps
+      percentage: 50
+```
+
+The configuration translates to the Envoy configuration in Http's filters:
+
+```yaml
+{
+"name": "envoy.filters.http.fault",
+"typed_config": {
+  "@type": "type.googleapis.com/envoy.extensions.filters.http.fault.v3.HTTPFault",
+  "delay": {
+  "fixed_delay": "5s",
+  "percentage": {
+    "numerator": 505000,
+    "denominator": "TEN_THOUSAND"
+  }
+  },
+  "abort": {
+  "http_status": 500,
+  "percentage": {
+    "numerator": 50
+  }
+  },
+  "headers": [
+  {
+    "name": "x-kuma-tags",
+    "safe_regex_match": {
+    "google_re2": {},
+    "regex": ".*&kuma.io/protocol=[^&]*http[,&].*&kuma.io/service=[^&]*frontend_default_svc_80[,&].*"
+    }
+  }
+  ],
+  "response_rate_limit": {
+  "fixed_limit": {
+    "limit_kbps": "50000"
+  },
+  "percentage": {
+    "numerator": 50
+  }
+  }
+}
+```
+
+### Specification
+
+Fault injection supports only `HTTP` protocol and is configured in `HttpFilter`. We can use the header `x-kuma-tags` to match the specific source. In the case of `MeshHealthCheck` which doesn't set `x-kuma-tags` we can put `x-kuma-tags` or create `MeshFaultInjection` that allows only `Mesh` in `From` section.
+
+#### Top level
+
+Top-level targetRef can have all available kinds:
+```yaml
+targetRef:
+ kind: Mesh|MeshSubset|MeshService|MeshServiceSubset
+ name: ...
+```
+Fault injection is an inbound policy so only `from` should be configured.
+
+#### From level
+
+We can allow configuration from different sources.
+```yaml
+from:
+ - targetRef:
+     kind: Mesh|MeshSubset|MeshService|MeshServiceSubset
+     name: ...
+```
+The only issue is `MeshHealthCheck` which doesn't set the `x-kuma-tags` flag and might not match the specific fault injection configuration. We have 2 options:
+
+* set `x-kuma-tags` header by default for `MeshHealthCheck`
+* change the scope of `From` to `Mesh` so the policy will affect all sources
+
+#### Configuration
+
+```yaml
+default:
```

Comment:
Technically it's possible to configure a list of HTTPFaults in Envoy. For example, you can configure Envoy to abort with 500, 503, and 504 status codes at the same time. But that's not possible with the existing FaultInjection policy. If we wanted to support this with MeshFaultInjection, that'd require us to convert `abort` into the list:
```yaml
default:
  abort:
    - httpStatus: 500
      percentage: 50
    - httpStatus: 503
      percentage: 50
    - httpStatus: 504
      percentage: 50
```
I don't know if we want to do this, because FaultInjection is not the most actively used policy and I doubt people will run into this use case. Just bringing it here since it's MADR we will discuss all possible options :)

Reply:
It doesn't seems like it allow the list: https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/http/fault/v3/fault.proto#envoy-v3-api-msg-extensions-filters-http-fault-v3-httpfault

Reply:
You can append many HTTPFaults as HTTPFilters https://github.com/kumahq/kuma/blob/master/pkg/xds/envoy/listeners/v3/fault_injection_configurer.go#L47

Reply:
That's correct but won't it be just another policy matching for another service? We set there header source so order is going to be important and probably first filter is going to be applied

Reply:
The use case here is having: 5% return 404 and another 5% return 500 no?

Reply:
Yes it might be this. I will add it. Probably this list should be dependable by other policies

Reply:
"another 5%" it's not another 5%. It's "the same" 5%.
```
abort:
- httpStatus: 500
   percentage: 50
- httpStatus: 503
   percentage: 50
```
Does not mean the request fail every single time with 5xx code.
That's why I think it's misleading.

I'd lean more towards
```
default:
  appendFaults:
  - abort:
      httpStatus: 500
      percentage: 50
  - abort:
      httpStatus: 500
      percentage: 10
```

which is 1:1 representation between elements and filter instances in Envoy

Reply:
```
default: 
  http:
    faults:
      - delay:
          percentage: 0.5
          value: "5s"
      - abort: 
           httpStatus: 500 
           percentage: 50 
       - abort: 
           httpStatus: 500
           percentage: 10
```